torch/distributed/fsdp/_unshard_param_utils.py

```diff
@@ -120,38 +124,49 @@ def _unflatten_as_params(state: _FSDPState, module: nn.Module) -> Generator:
                 _register_flat_param(state, module)
 
 
-@contextlib.contextmanager
-def _unshard_params(
-    module: nn.Module,
+def _validate_unshard_params_args(
     state: _FSDPState,
-    writeback: bool = True,
-    rank0_only: bool = False,
-    offload_to_cpu: bool = False,
-    with_grads: bool = False,
-):
+    writeback: bool,
+    rank0_only: bool,
+    offload_to_cpu: bool,
+    with_grads: bool,
+) -> None:
     if with_grads and (offload_to_cpu or not state._use_orig_params):
         raise NotImplementedError(
-            f"with_grads={with_grads} "
-            f"use_orig_params={state._use_orig_params} "
+            f"with_grads={with_grads}, "
+            f"use_orig_params={state._use_orig_params}, "
             f"offload_to_cpu={offload_to_cpu} "
             f"is not supported yet"
         )
     if writeback and rank0_only:
-        raise ValueError(
-            "writeback=True and rank0_only=True is not supported, as model "
-            "parameter shapes will be different across ranks, and writing "
-            "to them can lead to inconsistencies across ranks when the "
-            "context is exited."
-        )
+        # TODO: Rank 0 can broadcast the `FlatParameter` to allow all ranks to
```

Comment:
could we file an issue for this? would it work for use_orig_params=True as well?

Reply:
I think it should work for both `use_orig_params=True` and `False`. I will file an issue.

Reply:
https://github.com/pytorch/pytorch/issues/93938